src/core/service.c

```diff
@@ -4308,6 +4308,19 @@ static int service_can_clean(Unit *u, ExecCleanMask *ret) {
         return exec_context_get_clean_mask(&s->exec_context, ret);
 }
 
+static const char *service_finished_job(Unit *u, JobType t, JobResult result) {
+        if (t == JOB_START && result == JOB_DONE) {
+                Service *s = SERVICE(u);
+
+                if (s->type == SERVICE_ONESHOT)
+                        return "Finished %s.";
+                else
+                        return "Started %s.";
```

Comment:
Maybe remove those two lines, and instead add a comment below before `return NULL` that a fallback to the generic logic is made. Handling some of the cases in a way that duplicates the fallback logic and falling through to the fallback logic in other cases seems unclear. I'd prefer to only handle the special case here, and always return NULL if the fallback is good enough.

Reply:
Agreed; this branch removed, short comment before `return NULL` added.